src/posts/what-is-x-forwarded-for-about.md

```diff
@@ -0,0 +1,271 @@
+---
+title: 'What is X-Forwarded-For All About?'
+date: '2024-01-03T15:26'
+# cover_image: './header-images/notepad.jpg'
+author: Phil Sturgeon
+authorUrl: https://philsturgeon.com/
+---
+
+The X-Forwarded-For (XFF) HTTP header serves a crucial role in providing insight into the origin of web requests. The header works as a mechanism for conveying the original source IP addresses of clients, and not just one client, but multiple. This list of IPv4 and IPv6 addresses is handy in scenarios where requests traverse several servers, proxies, or load balancers.
+
+A typical HTTP request goes on a bit of a journey, traversing multiple layers of infrastructure before reaching its destination. Without the `X-Forwarded-For` header, the receiving server would only see the IP address of the last intermediary in the chain, rather than the true client origin. 
+
+<mermaid-js>
+sequenceDiagram
+    actor Client
+    participant CDN
+    participant LB as Load Balancer
+    participant B as Backend App X
+    Client->>CDN: 28.178.124.142
+    CDN->>LB: 198.40.10.101
+    LB->>B: 198.40.10.102
+</mermaid-js>
+
+By the time the backend application is seeing an incoming request, the IP address is long forgotten, which is where `X-Forwarded-By` can help out.
+
+```http
+X-Forwarded-By: 28.178.124.142, 198.40.10.101, 198.40.10.102
```

Comment:
Ah looking closer - I think the confusion maybe because this is using `X-Forwarded-By` instead of `X-Forwarded-For`? I think the former is very rarely used (it's often hard to know which IP should actually be used in practice) and we just want to focus on `-For` here.

Reply:
I dont know if -By even exists I just typed the wrong this, its meant to be For. 

Yeah the backend IP is 103 so the examples with 101 and 102 are correct.

Reply:
Hmm, no I don't think 102 should be here. Who adds that?

As far as I'm aware, nobody ever adds their own address to the header, they always add the downstream client's address, and then the last hop never adds any addresses at all. 

102 is the 2nd last hop here, as I understand it? That means only the server at the last hop sees that address, and the server doesn't modify the headers, so it never appears in XFF.

It's easier to see building up from simpler examples:

* Client -> server has no XFF header - there's no forwarding involved
* Client -> CDN -> server has 1 XFF header value - the CDN adds it with the client's IP
* Client -> CDN -> LB -> server has 2 XFF header values (CDN adds client's IP, LB adds CDN's ip) not 3

Reply:
102 is definitely meant to be there, but I think the diagram is not as clear as it could be. 

Does this help? 

<img width="608" alt="image" src="https://github.com/httptoolkit/httptoolkit-website/assets/67381/f1553289-2211-40c5-b4b8-3283bb544f65">

Reply:
> The nginx config of the backend app

Ahhh, I think this might be the issue here. I'm assuming there's no nginx configuration in the backend app, and it sounds like you're not.

I think no nginx here is a clearer model - if there was an nginx server, it would be a separate node in front of the backend app. I would assume the backend app node is a single process listening on a port (e.g. a Node server using express).

In the nginx case, note that it only ever adds this header when it's actually forwarding to a different server elsewhere, never when it's serving directly, because both `$proxy_add_x_forwarded_for` and `proxy_set_header` only apply to the nginx's proxy module. In the few scenarios where nginx is the last node (e.g. CGI scripts) the same logic would apply as I'm describing, where the IP of the node before the nginx server would not appear in the header.

> Any server could be confidured to add an IP or not add an IP, but it should do that

I've done more digging, and I'm pretty sure this isn't even desirable or occasional - 102 appearing in this header should never ever happen.

Primarily because a backend server that receives a request shouldn't actively change the incoming headers before the code within reads them. If it does, it's lying to the application logic within and you're asking for trouble. Only intermediaries should add to this header, not servers. Express.js should accurately tell you the headers it receives, without inventing extra values. If (as above) there is another implicit nginx in front that makes sense, but I think it's clearer to only include proxies like nginx in examples as explicit nodes.

[Wikipedia's article](https://en.wikipedia.org/wiki/X-Forwarded-For) on the topic backs this up, and specifically says:

> the last proxy IP address in a chain is not contained within the X-Forwarded-For field

Similarly the [forwarded spec](https://datatracker.ietf.org/doc/html/rfc7239#section-5.2) says:

> The last proxy in the chain is not part of the list of "for" parameters.  The last proxy's IP address, and optionally a port number, are, however, readily available as the remote IP address at the transport layer.

The header is not a list of all intermediaries. It's a list of all traffic sources that have been omitted from the transport layer (TCP header) due to forwarding. The node directly preceeding the server never appears.

Does that make more sense?

Given that, it's always wrong for 102 to appear here, and the same applies for all the other examples in this post. Let's assume there's no implicit nginx layer here - can you update these examples please when you have a minute?

Reply:
Aha! Brilliant, I had read those quotes as meaning something else entirely. 

Thanks for catching that. Two heads are better than one.

Article is updated.